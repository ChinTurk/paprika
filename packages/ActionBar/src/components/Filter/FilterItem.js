import React from "react";
import PropTypes from "prop-types";
import Button from "@paprika/button";
import useI18n from "@paprika/l10n/lib/useI18n";
import DatePicker from "./DatePicker";
import InlineSelect from "../InlineSelect/InlineSelect";
import InlineInput from "../InlineInput/InlineInput";
import FilterContext from "./context";
import * as types from "../../types";
import * as sc from "./Filter.styles";

const propTypes = {
  columnId: PropTypes.string.isRequired,
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  index: PropTypes.number.isRequired,
  onChange: PropTypes.func.isRequired,
  onDelete: PropTypes.func.isRequired,
  renderValueField: PropTypes.func,
  rule: PropTypes.string.isRequired,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]).isRequired,
  data: PropTypes.arrayOf(PropTypes.shape({})),
};

const defaultProps = {
  id: null,
  renderValueField: null,
  data: null,
};

function FilterItem(props) {
  const {
    index,
    onChange,
    onDelete,
    id,
    columnId: selectedColumnId,
    rule: selectedRule,
    value,
    renderValueField: renderCustomValueField,
    data,
  } = props;
  const { columns, filtersRef, onChangeOperator, operator, rulesByType } = React.useContext(FilterContext);
  const I18n = useI18n();

  const selectedColumnType = columns.find(({ id }) => id === selectedColumnId).type;

  const selectOptions = React.useMemo(() => {
    return selectedColumnType === types.COLUMN_SINGLE_SELECT
      ? data.filter((obj, index, arr) => {
          return arr.map(mapObj => mapObj[selectedColumnId]).indexOf(obj[selectedColumnId]) === index;
        })
      : null;
  }, [selectedColumnType, data, selectedColumnId]);

  function handleRemoveFilter() {
    filtersRef.current.focus();
    onDelete(id);
  }

  function handleChangeColumn(event) {
    const newColumnId = event.target.value;
    onChange(types.COLUMN, { id, columnId: newColumnId });
  }

  function handleChangeRule(event) {
    onChange(types.RULE, { id, rule: event.target.value });
  }

  function handleChangeValue(newValue) {
    onChange(types.FILTER_VALUE, { id, value: newValue });
  }

  function handleChangeDatePicker(momentDate) {
    filtersRef.current.focus();

    if (momentDate === null) {
      handleChangeValue("");
    } else {
      handleChangeValue(momentDate.format(columns.find(({ id }) => id === selectedColumnId).momentParsingFormat));
    }
  }

  function handleChangeBooleanFilterValue(event) {
    onChange(types.FILTER_VALUE, { id, value: event.target.value === "true" });
  }

  function handleChangeSingleSelectFilterValue(event) {
    onChange(types.FILTER_VALUE, { id, value: event.target.value });
  }

  function renderRuleField() {
    switch (selectedColumnType) {
      case types.COLUMN_BOOLEAN:
        return I18n.t("actionBar.filter.rules.is");
      default:
        return (
          <InlineSelect
            onChange={handleChangeRule}
            value={selectedRule}
            selectedLabel={I18n.t(`actionBar.filter.rules.${types.localeKeysByRule[selectedRule]}`)}
            data-pka-anchor="actionBar.filter.ruleSelector"
          >
            {rulesByType[selectedColumnType].map(rule => (
              <option key={rule} value={rule}>
                {I18n.t(`actionBar.filter.rules.${types.localeKeysByRule[rule]}`)}
              </option>
            ))}
          </InlineSelect>
        );
    }
  }
  function renderValueField() {
    const shouldNotShowValueField =
      selectedRule === types.IS_BLANK ||
      selectedRule === types.IS_NOT_BLANK ||
      selectedRule === types.IS_EMPTY ||
      selectedRule === types.IS_NOT_EMPTY;

    if (shouldNotShowValueField) return null;

    if (renderCustomValueField) return renderCustomValueField();

    switch (selectedColumnType) {
      case types.COLUMN_BOOLEAN:
        return (
          <InlineSelect
            value={`${value}`}
            onChange={handleChangeBooleanFilterValue}
            selectedLabel={I18n.t(`actionBar.filter.rules.${value}`)}
          >
            <option value="true">{I18n.t("actionBar.filter.rules.true")}</option>
            <option value="false">{I18n.t("actionBar.filter.rules.false")}</option>
          </InlineSelect>
        );
      case types.COLUMN_DATE:
        return (
          <DatePicker
            initialDate={value}
            onChange={handleChangeDatePicker}
            parsingFormat={columns.find(({ id }) => id === selectedColumnId).momentParsingFormat}
          />
        );

      case types.COLUMN_SINGLE_SELECT:
        return (
          <InlineSelect value={value} onChange={handleChangeSingleSelectFilterValue} selectedLabel={value}>
            {selectOptions.map(data => (
              <option key={data[selectedColumnId]} value={data[selectedColumnId]}>
                {data[selectedColumnId]}
              </option>
            ))}
          </InlineSelect>
        );
      default:
        return (
          <InlineInput inputKey={`${selectedColumnId}-${selectedRule}`} value={value} onChange={handleChangeValue} />
        );
    }
  }

  function renderPrefix() {
    const staticPrefix = I18n.t(`actionBar.filter.${operator === types.AND ? "and" : "or"}`);

    switch (index) {
      case 0:
        return I18n.t("actionBar.filter.where");
      case 1:
        if (onChangeOperator) {
          return (
            <InlineSelect
              onChange={onChangeOperator}
              value={operator}
              selectedLabel={operator === types.AND ? I18n.t("actionBar.filter.and") : I18n.t("actionBar.filter.or")}
            >
              <option value={types.AND}>{I18n.t("actionBar.filter.and")}</option>
              <option value={types.OR}>{I18n.t("actionBar.filter.or")}</option>
            </InlineSelect>
          );
        }
        return staticPrefix;
      default:
        return staticPrefix;
    }
  }
  return (
    <sc.FilterItem data-pka-anchor="actionBar.filter.filterItem">
      <Button.Close data-pka-anchor="actionBar.filter.deleteFilterButton" onClick={handleRemoveFilter} size="small" />
      {renderPrefix()}
      <InlineSelect
        onChange={handleChangeColumn}
        value={selectedColumnId}
        selectedLabel={columns.find(column => column.id === selectedColumnId).label}
      >
        {columns.map(column => (
          <option key={column.id} value={column.id}>
            {column.label}
          </option>
        ))}
      </InlineSelect>

      {renderRuleField()}
      {renderValueField()}
    </sc.FilterItem>
  );
}

FilterItem.propTypes = propTypes;
FilterItem.types = types;
FilterItem.defaultProps = defaultProps;

export default React.memo(FilterItem);
