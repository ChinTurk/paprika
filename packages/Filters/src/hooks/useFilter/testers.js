import moment from "moment";
import Filters from "../../Filters";

const is = (value, testValue) => (testValue === "" ? true : value === testValue);
const isNot = (value, testValue) => (testValue === "" ? true : value !== testValue);
const isBlank = value => value === null || value === undefined || value === "";
const isNotBlank = value => value !== null && value !== undefined && value !== "";
const processNumber = (value, testValue, testFunction) => {
  if (testValue === "") return true;
  const testNumber = parseFloat(testValue, 10);
  return Number.isNaN(testNumber) ? false : testFunction(value, testNumber);
};
const processDate = (momentParsingFormat, value, testValue, testFunction) => {
  if (testValue === "") return true;
  const testDate = moment(testValue, momentParsingFormat);
  if (testDate.isValid()) return testFunction(moment(value, momentParsingFormat), testDate);
  return false;
};

const testers = {
  [Filters.rules.IS]: is,
  [Filters.rules.IS_NOT]: isNot,
  [Filters.rules.CONTAINS]: (value, testValue) =>
    testValue === "" ? true : `${value}`.match(new RegExp(testValue, "i")),
  [Filters.rules.DOES_NOT_CONTAIN]: (value, testValue) =>
    testValue === "" ? true : !value.match(new RegExp(testValue, "i")),
  [Filters.rules.IS_BLANK]: isBlank,
  [Filters.rules.IS_NOT_BLANK]: isNotBlank,
  [Filters.rules.EQUALS]: (value, testValue) => processNumber(value, testValue, is),
  [Filters.rules.NOT_EQUAL_TO]: (value, testValue) => processNumber(value, testValue, isNot),
  [Filters.rules.GREATER_THAN]: (value, testValue) => processNumber(value, testValue, (a, b) => a > b),
  [Filters.rules.GREATER_THAN_OR_EQUAL_TO]: (value, testValue) => processNumber(value, testValue, (a, b) => a >= b),
  [Filters.rules.LESS_THAN]: (value, testValue) => processNumber(value, testValue, (a, b) => a < b),
  [Filters.rules.LESS_THAN_OR_EQUAL_TO]: (value, testValue) => processNumber(value, testValue, (a, b) => a <= b),
  [Filters.rules.IS_EMPTY]: isBlank,
  [Filters.rules.IS_NOT_EMPTY]: isNotBlank,
  [Filters.rules.IS_BEFORE]: (value, testValue, { momentParsingFormat }) =>
    processDate(momentParsingFormat, value, testValue, (a, b) => a.isBefore(b)),
  [Filters.rules.IS_AFTER]: (value, testValue, { momentParsingFormat }) =>
    processDate(momentParsingFormat, value, testValue, (a, b) => a.isAfter(b)),
};

export default testers;
