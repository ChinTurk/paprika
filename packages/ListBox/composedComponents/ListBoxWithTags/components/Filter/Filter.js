import React from "react";
import PropTypes from "prop-types";
import FilterStyled from "./Filter.styles";
import useStore from "../../../../store/useStore";
import * as actionTypes from "../../../../store/actionTypes";
import handleKeyboardKeys from "../../../../helpers/handleKeyboardKeys";

const propTypes = {
  placeholder: PropTypes.string.isRequired,
  defaultTagInputText: PropTypes.string,
};
const defaultProps = {
  defaultTagInputText: "",
};

export default function TagInput(props) {
  // const [inputText, setInputText] = React.useState(props.defaultTagInputText);
  const [state, dispatch] = useStore();
  // const handleChange = event => {
  //   setInputText(event.target.value);
  // };

  const handleClick = event => {
    if (!state.isPopoverOpen) {
      dispatch({ type: actionTypes.openPopover });
    }
    event.stopPropagation();
  };

  // this allowed the user navigate between the tags with [←] [→] left and right arrows

  const handleKeyDown = event => {
    handleKeyboardKeys(state, dispatch)(event);

    if (event.key === "Tab" && document.activeElement === state.refFilterInput.current) {
      dispatch({ type: actionTypes.closePopover });
    }
    // Avoid propagation of the cursor
    // at the start or at end of the input when the user click Arrow Up or Down
    // Also prevent to firing the space bar or enter key while filtering
    // if (event.key === "Backspace") {
    //   event.stopPropagation();
    // }

    //   event.preventDefault();
    //   event.stopPropagation();
    // }

    // if (event.key === " ") {
    //   event.stopPropagation();
    // }
    //

    // /////////////////////////////////

    // if (event.key === "ArrowUp" || event.key === "ArrowDown") {
    //   setActiveTag(null);
    //   if (!state.refFilterInput.current) {
    //     dispatch({ type: actionTypes.openPopover });
    //     return;
    //   }
    //
    //   state.refFilterInput.current.focus();
    //
    //   if (state.refFilterInput.current && (state.activeOption === 0 && event.key === "ArrowUp")) {
    //     state.refFilterInput.current.focus();
    //     setActiveTag(-1);
    //   }
    // }

    // if (state.selectedOptions.length) {
    //   if (event.key === "ArrowLeft" || event.key === "ArrowRight") {
    //     if (state.refFilterInput.current === document.activeElement) {
    //       state.refFilterInput.current.focus();
    //       setActiveTag(-1);
    //       return;
    //     }
    //
    //     if (activeTag === null) {
    //       setActiveTag(state.selectedOptions[0]);
    //
    //       return;
    //     }
    //
    //     // this means the focus is on the TagInput
    //     if (activeTag === -1) {
    //       if (event.key === "ArrowLeft") {
    //         setActiveTag(state.selectedOptions[state.selectedOptions.length - 1]);
    //       } else {
    //         setActiveTag(0);
    //       }
    //       state.refListBoxContainer.current.focus();
    //       return;
    //     }
    //
    //     const index = state.selectedOptions.indexOf(activeTag);
    //
    //     if (
    //       (state.refFilterInput.current && index === 0 && event.key === "ArrowLeft") ||
    //       (index === state.selectedOptions.length - 1 && event.key === "ArrowRight")
    //     ) {
    //       setActiveTag(-1);
    //       state.refFilterInput.current.focus();
    //       return;
    //     }
    //
    //     const currentActiveIndex = event.key === "ArrowLeft" ? index - 1 : index + 1;
    //     const nextIndex = (currentActiveIndex + state.selectedOptions.length) % state.selectedOptions.length;
    //
    //     setActiveTag(state.selectedOptions[nextIndex]);
    //     // state.refListBoxContainer.current.focus();
    //   }
    //
    //   if (event.key === "Backspace") {
    //     if (activeTag === null) {
    //       setActiveTag(state.selectedOptions[state.selectedOptions.length - 1]);
    //       return;
    //     }
    //
    //     if (state.selectedOptions.length - 1 >= 0) {
    //       dispatch({ type: actionTypes.unselectOptions, payload: [activeTag] });
    //       setActiveTag(state.selectedOptions[state.selectedOptions.length - 1]);
    //     } else {
    //       setActiveTag(null);
    //     }
    //   }
    // }
  };

  const handleBlur = () => {
    window.requestAnimationFrame(() => {
      if (
        !state.refTriggerContainer.current.contains(document.activeElement) &&
        state.refTriggerContainer.current !== document.activeElement &&
        !state.refListBoxContainer.current.contains(document.activeElement)
      ) {
        dispatch({ type: actionTypes.closePopover });
      }
    });
  };

  return (
    <FilterStyled
      placeholder={props.placeholder}
      onKeyDown={handleKeyDown}
      onBlur={handleBlur}
      onClick={handleClick}
      forceShowFilter
      defaultTextSearch={props.defaultTagInputText}
      hasSearchIcon={false}
    />
  );
}

TagInput.propTypes = propTypes;
TagInput.defaultProps = defaultProps;
