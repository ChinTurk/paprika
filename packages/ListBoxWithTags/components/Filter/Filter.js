import React from "react";
import PropTypes from "prop-types";
import useListBox from "@paprika/listbox/store/useListBox";
import handleKeyboardKeys from "@paprika/listbox/helpers/handleKeyboardKeys";
import FilterStyled from "./Filter.styles";

const propTypes = {
  defaultTagInputText: PropTypes.string,
  placeholder: PropTypes.string.isRequired,
  setActiveTag: PropTypes.func.isRequired,
  onAddCustomTag: PropTypes.func.isRequired,
};

const defaultProps = {
  defaultTagInputText: "",
};

export default function TagInput(props) {
  const [state, dispatch] = useListBox();
  const [filterText, setFilterText] = React.useState(props.defaultTagInputText);
  function openPopover() {
    dispatch({
      type: useListBox.types.openPopover,
    });
  }

  function closePopover() {
    dispatch({ type: useListBox.types.closePopover });
  }

  const handleClick = event => {
    if (!state.isPopoverOpen) {
      openPopover();
    }
    event.stopPropagation();
  };

  // this allowed the user navigate between the tags with [←] [→] left and right arrows

  const handlePressEnterKey = event => {
    if (event.target.value) {
      props.onAddCustomTag(state, event.target.value);
      setFilterText("");
    }
  };

  const handleKeyDown = event => {
    if (!state.isPopoverOpen) {
      openPopover();
    }

    if (event.key === "Enter") {
      if (state.hasNoResults) {
        handlePressEnterKey(event);
        return;
      }
    }

    if (event.key === " ") {
      event.stopPropagation();
      return;
    }

    if (event.key === "Backspace" && event.target.value.length) {
      event.stopPropagation();
    }

    if ((event.key === "ArrowLeft" || event.key === "ArrowRight") && event.target.value.length) {
      event.stopPropagation();
    }

    if (event.key === "Tab") {
      closePopover();
    }

    handleKeyboardKeys(state, dispatch)(event);
  };

  const handleBlur = () => {
    window.requestAnimationFrame(() => {
      if (document.activeElement !== state.refListBoxContainer.current) {
        closePopover();
        props.setActiveTag(null);
      }
    });
  };

  const handleFilterChange = event => {
    if (event.target.value.length) {
      props.setActiveTag(null);
    }

    setFilterText(event.target.value);
  };

  React.useLayoutEffect(() => {
    if (state.isPopoverOpen) {
      state.refFilterInput.current.focus();
      setFilterText("");
    }
  }, [state.selectedOptions]);

  return (
    <FilterStyled
      placeholder={props.placeholder}
      onKeyDown={handleKeyDown}
      onBlur={handleBlur}
      onClick={handleClick}
      onChangeFilter={handleFilterChange}
      forceShowFilter
      value={filterText}
      hasSearchIcon={false}
    />
  );
}

TagInput.propTypes = propTypes;
TagInput.defaultProps = defaultProps;
