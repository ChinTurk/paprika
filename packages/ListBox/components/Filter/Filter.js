import React from "react";
import PropTypes from "prop-types";
import useListBox from "../../store/useListBox";
import { isOptionVisible } from "../../helpers/options";
import * as effects from "./effects";
import * as actionTypes from "../../store/actionTypes";

import { FilterContainerStyled, FilterInputStyled, FilterSearchIconStyled } from "./Filter.styles";

const propTypes = {
  filter: PropTypes.func,
  forceShowFilter: PropTypes.bool,
  onKeyDown: PropTypes.func,
  hasSearchIcon: PropTypes.bool,
  onChangeFilter: PropTypes.func,
  placeholder: PropTypes.string,
  renderFilter: PropTypes.func,
  value: PropTypes.string,
};

const defaultProps = {
  filter: null,
  forceShowFilter: false,
  onKeyDown: null,
  hasSearchIcon: true,
  onChangeFilter: null,
  placeholder: "Filter...",
  renderFilter: null,
  value: null,
};

export default function Filter(props) {
  const [state, dispatch] = useListBox();

  const [textSearch, setTextSearch] = React.useState(props.value);

  // this might be better using _.escapeRegExp by lodash. But good enough for now
  function escapeRegExp(str) {
    return str.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"); // $& means the whole matched string
  }

  function applyFilter({ filteredOptions, hasNoResults }) {
    dispatch({
      type: actionTypes.applyFilter,
      payload: {
        filteredOptions,
        hasNoResults,
      },
    });
  }

  function filter(textSearchValue) {
    const { options } = state;
    const keys = Object.keys(options);

    if (keys.length) {
      if (props.filter) {
        return props.filter(textSearchValue, options);
      }

      const filteredOptions = keys.filter(key => {
        const hasLabel = typeof options[key].content === "string" || options[key].label || null;

        if (hasLabel) {
          const label = options[key].content === "string" || options[key].label;

          const filterRegExp = new RegExp(escapeRegExp(textSearchValue), "gi");
          return label.match(filterRegExp);
        }

        throw new Error(
          `Textsearch: ${textSearchValue} during ${options[key]}.
          ListBox.Filter  filter: <ListBox.Option /> need to have
          a string as a children or please provide a label prop
          <ListBox.Option label='yourOptionDescription' />.`
        );
      });

      if (!filteredOptions.length) {
        return [];
      }

      return filteredOptions.map(key => Number.parseInt(key, 10)).filter(keyInt => isOptionVisible(state, keyInt));
    }

    return [];
  }

  const handleChangeFilter = event => {
    const textSearchValue = event.target.value;

    if (props.onChangeFilter) {
      props.onChangeFilter(event);
    } else {
      setTextSearch(textSearchValue);
    }

    const filteredOptions = filter(textSearchValue);
    const hasNoResults = textSearchValue && filteredOptions.length === 0;
    applyFilter({ filteredOptions, hasNoResults });
  };

  const handleKeyDown = event => {
    // Avoid propagation of the cursor
    // at the start or at end of the input when the user click Arrow Up or Down
    // Also prevent to firing the space bar or enter key while filtering

    if (event.key === "ArrowUp" || event.key === "ArrowDown") {
      event.preventDefault();
    }

    if (event.key === " ") {
      event.stopPropagation();
    }
  };

  const handleBlur = () => {
    window.requestAnimationFrame(() => {
      if (state.Footer) {
        // if has footer the responsible of closing is the Footer no the onblur event
        return;
      }

      if (document.activeElement !== state.refListBoxContainer.current) {
        // this will reset the activeOption and close the Popover
        dispatch({
          type: actionTypes.setActiveOption,
          payload: {
            activeOptionIndex: null,
            isPopoverOpen: false,
          },
        });
      }
    });
  };

  const handleEffectValue = effects.handleEffectValue(props, applyFilter);
  const handleEffectIsPopOverOpen = effects.handleEffectIsPopOverOpen(state, setTextSearch);
  const handleEffectTextSearch = effects.handleEffectTextSearch(textSearch, applyFilter);

  React.useEffect(handleEffectValue, [props.value]);
  React.useEffect(handleEffectIsPopOverOpen, [state.isPopoverOpen]);
  React.useEffect(handleEffectTextSearch, [textSearch]);

  if (props.forceShowFilter || (state.isInlineDisplay && state.hasFilter) || (state.hasFilter && state.isPopoverOpen)) {
    const { renderFilter, placeholder, value, onChangeFilter, ...moreProps } = props;
    if (renderFilter) {
      return props.renderFilter(props);
    }

    return (
      <FilterContainerStyled>
        {props.hasSearchIcon ? <FilterSearchIconStyled /> : null}
        <FilterInputStyled
          ref={state.refFilterInput}
          type="text"
          onChange={handleChangeFilter}
          onKeyDown={props.onKeyDown || handleKeyDown}
          onBlur={handleBlur}
          value={value || textSearch || ""}
          placeholder={placeholder}
          {...moreProps}
        />
      </FilterContainerStyled>
    );
  }

  return null;
}

Filter.propTypes = propTypes;
Filter.defaultProps = defaultProps;
