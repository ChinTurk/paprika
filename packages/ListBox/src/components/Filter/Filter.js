import React from "react";
import PropTypes from "prop-types";
import useListBox from "../../useListBox";
import * as effects from "./effects";

import { FilterContainerStyled, FilterInputStyled, FilterSearchIconStyled } from "./Filter.styles";

const propTypes = {
  filter: PropTypes.func,
  filterExcludeSelectedOptions: PropTypes.bool,
  forceShowFilter: PropTypes.bool,
  hasSearchIcon: PropTypes.bool,
  onChangeFilter: PropTypes.func,
  onKeyDown: PropTypes.func,
  placeholder: PropTypes.string,
  renderFilter: PropTypes.func,
  value: PropTypes.string,
};

const defaultProps = {
  filter: null,
  filterExcludeSelectedOptions: false,
  forceShowFilter: false,
  hasSearchIcon: true,
  onChangeFilter: null,
  onKeyDown: null,
  placeholder: "Filter...",
  renderFilter: null,
  value: null,
};

export default function Filter(props) {
  const [state, dispatch] = useListBox();

  const [textSearch, setTextSearch] = React.useState(props.value);

  // this might be better using _.escapeRegExp by lodash. But good enough for now
  function escapeRegExp(str) {
    return str.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"); // $& means the whole matched string
  }

  function applyFilter({ filteredOptions, noResultsFound }) {
    dispatch({
      type: useListBox.types.applyFilter,
      payload: {
        filteredOptions,
        noResultsFound,
      },
    });
  }

  function filter(textSearchValue) {
    const { options } = state;
    const keys = Object.keys(options);

    if (keys.length) {
      const filteredOptions = keys.filter(key => {
        const hasLabel = typeof options[key].content === "string" || options[key].label || null;

        if (hasLabel) {
          const label = options[key].content === "string" || options[key].label;

          const filterRegExp = new RegExp(escapeRegExp(textSearchValue), "gi");
          return label.match(filterRegExp);
        }

        throw new Error(
          `Textsearch: ${textSearchValue} during ${options[key]}.
          ListBox.Filter  filter: <ListBox.Option /> need to have
          a string as a children or please provide a label prop
          <ListBox.Option label='yourOptionDescription' />.`
        );
      });

      if (!filteredOptions.length) {
        return [];
      }

      if (props.filterExcludeSelectedOptions) {
        return filteredOptions
          .map(key => Number.parseInt(key, 10))
          .filter(keyInt => !state.selectedOptions.includes(keyInt));
      }

      return filteredOptions.map(key => Number.parseInt(key, 10));
    }

    return [];
  }

  const handleChangeFilter = event => {
    const textSearchValue = event.target.value;

    if (state.isDisabled) return;

    if (props.filter) {
      setTextSearch(textSearchValue);
      return props.filter({ search: textSearchValue, state, dispatch });
    }

    if (props.onChangeFilter) {
      props.onChangeFilter(event);
    } else {
      setTextSearch(textSearchValue);
    }

    const filteredOptions = filter(textSearchValue);
    const noResultsFound = textSearchValue && filteredOptions.length === 0;
    applyFilter({ filteredOptions, noResultsFound });
  };

  const handleBlur = () => {
    window.requestAnimationFrame(() => {
      if (state.hasFooter) {
        // if has footer the responsible of closing is the Footer no the onblur event
        return;
      }

      if (document.activeElement !== state.refListBoxContainer.current) {
        // this will reset the activeOption and close the Popover
        dispatch({
          type: useListBox.types.setActiveOption,
          payload: {
            activeOptionIndex: null,
          },
        });
      }
    });
  };

  const handleEffectValue = effects.handleEffectValue(props, applyFilter);
  const handleEffectIsPopOverOpen = effects.handleEffectIsPopOverOpen(state, setTextSearch);
  const handleEffectTextSearch = effects.handleEffectTextSearch(textSearch, applyFilter);

  React.useEffect(handleEffectValue, [props.value]);
  React.useEffect(handleEffectIsPopOverOpen, [state.isPopoverOpen]);
  React.useEffect(handleEffectTextSearch, [textSearch]);

  if (props.forceShowFilter || (state.isInline && state.hasFilter) || (state.hasFilter && state.isPopoverOpen)) {
    const { renderFilter, placeholder, value, onChangeFilter, filter, ...moreProps } = props;
    if (renderFilter) {
      return props.renderFilter(props);
    }

    return (
      <FilterContainerStyled>
        {props.hasSearchIcon ? <FilterSearchIconStyled /> : null}
        <FilterInputStyled
          isDisabled={state.isDisabled}
          onBlur={handleBlur}
          onChange={handleChangeFilter}
          onKeyDown={props.onKeyDown}
          placeholder={placeholder}
          ref={state.refFilterInput}
          type="text"
          value={value || textSearch || ""}
          {...moreProps}
        />
      </FilterContainerStyled>
    );
  }

  return null;
}

Filter.propTypes = propTypes;
Filter.defaultProps = defaultProps;
