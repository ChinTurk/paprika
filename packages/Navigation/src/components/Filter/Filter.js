import React from "react";
import PropTypes from "prop-types";
import Popover from "@paprika/popover";
import Button from "@paprika/button";
import useI18n from "@paprika/l10n/lib/useI18n";
import FilterItem from "./FilterItem";
import { rulesByType } from "./rules";
import * as styled from "./Filter.styles";

const propTypes = {
  onChange: PropTypes.func.isRequired,
  filters: PropTypes.arrayOf(
    PropTypes.shape({
      filterId: PropTypes.string,
      columnId: PropTypes.string.isRequired,
      rule: PropTypes.string.isRequired,
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    })
  ).isRequired,
  columns: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
    })
  ).isRequired,
  onDeleteFilter: PropTypes.func.isRequired,
  onAddFilter: PropTypes.func.isRequired,
};

export default function Filter(props) {
  const { onDeleteFilter, onAddFilter, filters, columns, onChange } = props;
  const I18n = useI18n();
  const filtersRef = React.useRef(null);

  function getLabelText(numberOfFilters) {
    switch (numberOfFilters) {
      case 0:
        return I18n.t("navigation.filter.label");
      case 1:
        return I18n.t("navigation.filter.singular_label");
      default:
        return I18n.t("navigation.filter.plural_label", { numberOfFilters });
    }
  }

  return (
    <Popover align="bottom" edge="left" maxWidth={600} offset={8}>
      <Popover.Trigger kind="flat">
        {(handler, attributes) => (
          <styled.Trigger {...attributes} isSemantic={false} onClick={handler} hasFilterApplied={filters.length > 0}>
            <styled.Icon />
            {getLabelText(filters.length)}
          </styled.Trigger>
        )}
      </Popover.Trigger>
      <Popover.Content>
        <Popover.Card>
          <styled.FiltersPanel ref={filtersRef} tabIndex={-1}>
            {filters.map((filter, index) => (
              <FilterItem
                key={filter.filterId}
                filter={filter}
                columns={columns}
                onDeleteFilter={onDeleteFilter}
                onChange={onChange}
                isFirst={index === 0}
                filtersRef={filtersRef}
              />
            ))}
          </styled.FiltersPanel>
          <Button onClick={onAddFilter} kind="minor">
            {I18n.t(`navigation.filter.add_filter`)}
          </Button>
        </Popover.Card>
        <Popover.Tip />
      </Popover.Content>
    </Popover>
  );
}

Filter.propTypes = propTypes;
Filter.displayName = "Navigation.Filter";
Filter.rulesByType = rulesByType;
