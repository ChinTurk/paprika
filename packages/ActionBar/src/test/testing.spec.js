import testers from "../hooks/useFilter/testers";
import Filter from "../components/Filter/Filter";

describe("action bar", () => {
  it.only("should just run", () => {
    expect(testers[Filter.rules.IS]("any value", "3")).toMatchSnapshot();
  });
});
