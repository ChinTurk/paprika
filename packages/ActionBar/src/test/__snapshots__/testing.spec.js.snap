// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action bar should just run 1`] = `true`;
